README



       JGroups - A Framework for Group Communication in Java
       ========================================================

			    March 3, 1998

			       Bela Ban
			   4114 Upson Hall
			  Cornell University
			   Ithaca, NY 14853
			  bba@cs.cornell.edu
		       belaban@gmail.com


JGroups is a Java library for reliable group communication. It
consists of 3 parts: (1) a socket-like API for application
development, (2) a protocol stack, which implements reliable
communication, and (3) a set of building blocks, which give the
developer high-level abstractions (e.g. ReplicatedHashMap, an
implementation of java.util.Map).

The API (a channel) looks like a socket: there are methods for joining
and leaving groups, sending and receiving messages,
getting the shared group state, and registering for notifications when
a member joins, or an existing member leaves or crashes.

The protocol stack is a list of protocols, through which each
message passes. Each protocol implements an up() and down()
method, and may modify, reorder, encrypt, fragment/unfragment, drop,
or pass a message up/down. The protocol stack is created
according to a specification given when a channel is created. New
protocols can be plugged into the stack easily.

Building blocks hide the channel and provide a higher abstraction.
Example: ReplicatedHashMap implements java.util.Map and implements
all methods that change the map (clear(), put(), remove()).
Those methods are invoked on all hashmap instances in the same group
simultaneously, so that all hashmaps have the same state.
A new hashmap uses a state transfer protocol to initially obtain
the shared group state from an existing member. This allows for 
replication of data structures across processes.



Group communication is important in the following situations:

 - A service has to be replicated for availability. As long as at
   least one of the servers remains operational, the service itself
   remains operational

 - Service requests have to be balanced between a set of servers

 - A large number of objects have to be managed as one entity (e.g. a
   management domain)

 - Notification service / push technology: receivers subscribe to a
   channel, senders send data to the channels, channels distribute
   data to all receivers subscribed to the channel.
   Used for example for video distribution, videoconferencing
	
	
JGroups deliberately models a rather low-level message-oriented
middleware (MOM) model. The reason is that we don't want to impose a
one-size-fits-all model on the programmer, who usually will want to
extend the model in various (previously unconceived) ways anyway.

Providing low level Java classes allows the programmer to
extend/replace classes at will, as the granularity of the system is
finer.






doc/history.txt


			     History List
			     ============


[For current version, see file Version.java (or invoke 'java org.jgroups.Version')]


bba         = Bela Ban, bba@cs.cornell.edu
bela        = Bela Ban, belaban@yahoo.com
i-scream    = Gianluca Collot, gianlucac@tin.it
igeorg      = John Georgiadis, i.georgiadis@doc.ic.ac.uk
jmenard     = Jim Menard (jimm@io.com)
fhanik      = Filip Hanik (filip@filip.net)
vlada       = Vladimir Blagojevic (vladimir@cs.yorku.ca)
rrokytskyy  = Roman Rokytskyy (rrokytskyy@acm.org)
akbollu     = Ananda Bollu (akbollu@users.sf.net)
whizkid_bay = Mandar Shinde (whizkid_bay@users.sf.net)
ovidiuf     = Ovidiu Feodorov (ovidiuf@users.sf.net)
romuald     = Romuald du Song
yaronr		= Yaron Rosenbaum (yaronr@mercury.co.il)
publicnmi	= Robert Schaffar-Taurok (robert@fusion.at)
ossiejnr    = Chris Mills (chris.mills@jboss.com)






***********
THIS FILE IS NOT UPDATED ANYMORE AS OF MAY 2 2006, I SWITCHED TO JIRA FOR THE ROADMAP AND HISTORY.
https://issues.redhat.com/browse/JGRP
***********





Version 2.3
-----------

- Changed method signature of RpcDispatcher.callRemoteMethod() to throw a Throwable.
  Previously it returned the exception as an object, now the exception will be thrown.
  Callers of these methods have to change their code, so this is an incompatible change. However,
  these calls are not used in JBossCache and JBoss Clustering.
  (https://issues.redhat.com/browse/JGRP-154)
  (bela Feb 16 2006)

- Added encryption of entire message to ENCRYPT
  (https://issues.redhat.com/browse/JGRP-190)
  (bela Feb 9 2006)

- Converted Word prog guide to docbook (./doc/progguide). Done by gdvieira@ic.unicamp.br
  (gdvieira@ic.unicamp.br Jan 26 2006)

- Created global JGroups thread group (all threads belong to it)
  (bela Jan 19 2006)

- Added AUTH protocol, samle configs and documentation. https://issues.redhat.com/browse/JGRP-164
  (ossiejnr Jan 12 2006)

- Fix in ReplicatedHashtable where state transfer notification is not emitted
  (https://issues.redhat.com/browse/JGRP-175)
  (David Forget Jan 5 2006)

- init() is now called after all protocols have been created
  (bela Jan 5 2006)

- Fixed bug where srv_sock_bind_addr could be different from transport's bind_addr.
  This could cause incorrect suspect messages.
  (https://issues.redhat.com/browse/JGRP-173)
  (bela Jan 3 2006)

- Added total order SEQUENCER protocol. See doc/SEQUENCER.txt for details
  (bela Dec 30 2005)





Version 2.2.9.1
---------------

- Created label JGROUPS_2_2_9_1
  (bela Dec 29 2005)

- Fixed bug "binding to same interface twice fails"
  (https://issues.redhat.com/browse/JGRP-167)
  (bela Dec 23 2005)

- Fixed merge bug occurring when members are joining during a merge
  (https://issues.redhat.com/browse/JGRP-139)
  (bela Dec 23 2005)

- Fixed ENCRYPT bug where down messages were queued unnecessarily
  (JIRA: https://issues.redhat.com/browse/JGRP-166)
  (bela Dec 21 2005)

- Replaced Vector for members and pingable_members in FD with CopyOnWriteArrayList.
  Fixes  (https://issues.redhat.com/browse/JGRP-161)
  (bela Dec 16 2005)




Version 2.2.9
-------------

- Created label JGROUPS_2_2_9
  (bela Dec 9 2005)

- Rewrite of TCP_NIO
  (Scott Marlow and Alex Fu Nov 2005)

- Added support for Receiver in JChannel (push based message reception)
  (bela Oct 2005)

- Added JChannel.dumpStats(): returns information about the various protocols, and the channel itself,
  as a map. Currenty, only NAKACK, TP and FC have implementations
  (bela July 26 2005)

- Eliminated creation of 3 input and 2 output streams *per message* in TP. These streams are now
  created at TP startup and simply reset when a message is to be received or sent
  (bela July 25 2005) 

- Improved performance of RpcDispatcher, MethodCall (is now Stremable), added unit tests
  (bela July 25 2005)

- Improved javadoc: copied portions of User's Guide to doc comments in source files,
  fixed javadoc warnings; added project and package overviews; tweaked build target. 
  (chrislott July 17 2005)
  
- Headers are now always created in a Message, because we always add at least 1 header
  to a message
  (bela July 15 2005)

- Changed org.jgroups.util.Digest to use a HashMap rather than arrays
  (bela July 12 2005)

- Refactored UDP and TCP into extending a common transport (TP)
  (bela July 4 2005)

- Completed first version of JMX instrumentation
  (bela June 14 2005)

- Added STATS protocol, provides stats via JMX
  (bela June 7 2005)

- Added org.jgroups.jmx package, contains various adapters to expose channel and protocols
  via JMX. Added jmxri.jar (can be removed once JDK 5 is baseline)
  (bela June 1 2005)

- Replaced System.err.println() with log.error() in setProperties() of all protocols
  (bela May 30 2005)

- Added xmit_from_random_member: retransmits will go to a random member, rather than the
  original sender of the message. This eases the burden on senders in large groups
  (bela May 25 2005)

- Fixed https://issues.redhat.com/browse/JGRP-79 (problems with nulling src addresses on
  loopback adapter in Windows). See JGroups/docs/NullingSrcAddresses.txt for details.
  (bela May 19 2005)

- Fixed problem with PingWaiter/PingSender in applets (https://issues.redhat.com/browse/JGRP-86)
  (bela Ma 19 2005)

- Fixed STATE_TRANSFER stuck in Channel.queue bug (https://issues.redhat.com/browse/JGRP-80)
  (bela May 10 2005)

- Added support for multiple locked locks to the DistributedLockManager.
  (publicnmi Jun 08 2005)



Version 2.2.8
-------------

- Created branch JGROUPS_2_2_8
  (bela April 29 2005)

- Fixed problem with deadlock detection (e.g. in RpcDispatcher). Caller always passed new call stack,
  rather than adding to existing call stack
  (bela April 25 2005)

- Removed xerces JARs
  (bela April 25 2005)

- UDP: fixed incorrect marshalling of IpAddresses with additional_data
  (https://issues.redhat.com/browse/JGRP-63)
  (bela April 23 2005)


src/org/jgroups/Event.java
package org.jgroups;


/**
 * Event is a JGroups internal class used for inter-stack and intra-stack communication. 
 * 
 * @since 2.0
 * @author Bela Ban
 */
public class Event {
    public static final int CONNECT                            =  2;  // arg = cluster name (string)
    public static final int DISCONNECT                         =  4;  // arg = null (local address)
    public static final int VIEW_CHANGE                        =  6;  // arg = View (or MergeView in case of merge)
    public static final int SUSPECT                            =  9;  // arg = Collection<Address> (suspected members)
    public static final int BLOCK                              = 10;  // arg = null (used by FLUSH)
    public static final int FIND_MBRS                          = 11;  // arg = List<Address> (can be null) -> Responses
    public static final int FIND_INITIAL_MBRS                  = 12;  // timeout (ms) = null -> Responses
    public static final int FIND_MBRS_ASYNC                    = 13;  // arg = Consumer<PingData>
    public static final int MERGE                              = 14;  // arg = Map<Address,View>
    public static final int TMP_VIEW                           = 15;  // arg = View
    public static final int BECOME_SERVER                      = 16;  // sent when client has joined group
    public static final int GET_APPLSTATE                      = 17;  // get state from appl (arg=StateTransferInfo)
    public static final int GET_STATE                          = 19;  // arg = StateTransferInfo
    public static final int GET_STATE_OK                       = 20;  // arg = StateTransferInfo
    public static final int STABLE                             = 30;  // arg = long[] (stable seqnos for mbrs)
    public static final int GET_DIGEST                         = 39;  // arg= address (or null)
    public static final int SET_DIGEST                         = 41;  // arg = Digest
    public static final int OVERWRITE_DIGEST                   = 42;  // arg = Digest
    public static final int UNSUSPECT                          = 51;  // arg = Address (of unsuspected member)
    public static final int MERGE_DIGEST                       = 53;  // arg = Digest
    public static final int CONFIG                             = 56;  // arg = Map<String,Object> (config properties)
    public static final int SUSPEND_STABLE                     = 65;  // arg = Long (max_suspend_time)
    public static final int RESUME_STABLE                      = 66;  // arg = null
    public static final int SUSPEND					           = 68;  // arg = List<Address> (used by FLUSH)
    public static final int RESUME					           = 70;  // arg = null (used by FLUSH)
    public static final int STATE_TRANSFER_INPUTSTREAM         = 71;  // arg = InputStream
    public static final int STATE_TRANSFER_OUTPUTSTREAM        = 72;  // arg = OutputStream
    public static final int STATE_TRANSFER_INPUTSTREAM_CLOSED  = 73;  // arg = StateTransferResult
    public static final int UNBLOCK                            = 75;  // arg = null (indicate end of flush round)
    public static final int CLOSE_BARRIER                      = 76;  // arg = null
    public static final int OPEN_BARRIER                       = 77;  // arg = null
    public static final int REBROADCAST				           = 78;  // arg = Digest
    public static final int CONNECT_WITH_STATE_TRANSFER        = 80;  // arg = cluster name (string)
    public static final int GET_PHYSICAL_ADDRESS               = 87;  // arg = Address --> PhysicalAddress
    public static final int GET_LOGICAL_PHYSICAL_MAPPINGS      = 88;  // arg = boolean --> Map<Address,PhysicalAddress>
    public static final int ADD_PHYSICAL_ADDRESS               = 89;  // arg = Tuple<Address,PhysicalAddress> --> boolean
    public static final int REMOVE_ADDRESS                     = 90;  // arg = Address
    public static final int GET_LOCAL_ADDRESS                  = 91;  // arg = null --> UUID (local_addr)
    public static final int CONNECT_USE_FLUSH			       = 92;
    public static final int CONNECT_WITH_STATE_TRANSFER_USE_FLUSH = 93;
    public static final int SUSPEND_BUT_FAIL                      = 94; // used in FLUSH testing, no args
    public static final int LOCK                               = 95; // arg = LockInfo
    public static final int UNLOCK                             = 96; // arg = LockInfo
    public static final int UNLOCK_ALL                         = 97; // arg = null
    public static final int LOCK_AWAIT                         = 98; // arg = LockInfo
    public static final int LOCK_SIGNAL                        = 99; // arg = AwaitInfo
    public static final int IS_MERGE_IN_PROGRESS               = 100; // returns true or false
    public static final int GET_PHYSICAL_ADDRESSES             = 102; // arg = null (returns all physical addresses)
    public static final int SITE_UNREACHABLE                   = 104; // arg = SiteMaster (RELAY2)
    public static final int PUNCH_HOLE                         = 106; // arg = Address (member)
    public static final int CLOSE_HOLE                         = 107; // arg = Address (member)
    public static final int GET_VIEW_FROM_COORD                = 108;
    public static final int GET_PING_DATA                      = 109; // arg = cluster_name
    public static final int GET_SECRET_KEY                     = 111; // arg = null -> Tuple<SecretKey,byte[]> // PK+version
    public static final int SET_SECRET_KEY                     = 112; // arg = Tuple<SecretKey,byte[]> // PK+version
    public static final int UNLOCK_FORCE                       = 113; // arg = lock name
    public static final int INSTALL_MERGE_VIEW                 = 114; // arg = MergeView

    public static final int USER_DEFINED                       = 1000; // arg = <user def., e.g. evt type + data>



    public static final Event GET_DIGEST_EVT=new Event(Event.GET_DIGEST);


    private final int    type;       // type of event
    private final Object arg;        // must be serializable if used for inter-stack communication


    public Event(int type) {
        this.type=type;
        this.arg=null;
    }

    public Event(int type, Object arg) {
        this.type=type;
        this.arg=arg;
    }

    public int                  type()    {return type;}
    public int                  getType() {return type;}
    public <T extends Object> T arg()     {return (T)arg;}
    public <T extends Object> T getArg()  {return (T)arg;}



    public static String type2String(int t) {
        switch(t) {
            case CONNECT:	             return "CONNECT";
            case DISCONNECT:	         return "DISCONNECT";
            case VIEW_CHANGE:	         return "VIEW_CHANGE";
            case SUSPECT:                return "SUSPECT";
            case BLOCK:	                 return "BLOCK";
            case FIND_MBRS:              return "FIND_MBRS";
            case FIND_INITIAL_MBRS:	     return "FIND_INITIAL_MBRS";
            case FIND_MBRS_ASYNC:        return "FIND_MBRS_ASYNC";
            case TMP_VIEW:	             return "TMP_VIEW";
            case BECOME_SERVER:	         return "BECOME_SERVER";
            case GET_APPLSTATE:          return "GET_APPLSTATE";
            case GET_STATE:              return "GET_STATE";
            case GET_STATE_OK:           return "GET_STATE_OK";
            case STABLE:                 return "STABLE";
            case GET_DIGEST:             return "GET_DIGEST";
            case SET_DIGEST:             return "SET_DIGEST";
            case OVERWRITE_DIGEST:       return "OVERWRITE_DIGEST";
            case MERGE:                  return "MERGE";
            case UNSUSPECT:              return "UNSUSPECT";
            case MERGE_DIGEST:           return "MERGE_DIGEST";
            case CONFIG:                 return "CONFIG";
            case SUSPEND_STABLE:         return "SUSPEND_STABLE";
            case RESUME_STABLE:          return "RESUME_STABLE";
            case SUSPEND:        		 return "SUSPEND";
            case SUSPEND_BUT_FAIL:       return "SUSPEND_BUT_FAIL";
            case RESUME:     			 return "RESUME";
            case STATE_TRANSFER_INPUTSTREAM: return "STATE_TRANSFER_INPUTSTREAM";
            case STATE_TRANSFER_OUTPUTSTREAM:return "STATE_TRANSFER_OUTPUTSTREAM";
            case STATE_TRANSFER_INPUTSTREAM_CLOSED: return "STATE_TRANSFER_INPUTSTREAM_CLOSED";
            case UNBLOCK:                return "UNBLOCK";
            case CLOSE_BARRIER:          return "CLOSE_BARRIER";
            case OPEN_BARRIER:           return "OPEN_BARRIER";
            case REBROADCAST:            return "REBROADCAST";
            case CONNECT_WITH_STATE_TRANSFER:    return "CONNECT_WITH_STATE_TRANSFER";
            case GET_PHYSICAL_ADDRESS:   return "GET_PHYSICAL_ADDRESS";
            case GET_LOGICAL_PHYSICAL_MAPPINGS: return "GET_LOGICAL_PHYSICAL_MAPPINGS";
            case ADD_PHYSICAL_ADDRESS:   return "ADD_PHYSICAL_ADDRESS";
            case REMOVE_ADDRESS:         return "REMOVE_ADDRESS";
            case GET_LOCAL_ADDRESS:      return "GET_LOCAL_ADDRESS";
            case CONNECT_USE_FLUSH: return "CONNECT_USE_FLUSH";
            case CONNECT_WITH_STATE_TRANSFER_USE_FLUSH: return "CONNECT_WITH_STATE_TRANSFER_USE_FLUSH";
            case LOCK:                   return "LOCK";
            case UNLOCK:                 return "UNLOCK";
            case UNLOCK_ALL:             return "UNLOCK_ALL";
            case LOCK_AWAIT:             return "LOCK_AWAIT";
            case LOCK_SIGNAL:            return "LOCK_SIGNAL";
            case IS_MERGE_IN_PROGRESS:   return "IS_MERGE_IN_PROGRESS";
            case GET_PHYSICAL_ADDRESSES: return "GET_PHYSICAL_ADDRESSES";
            case SITE_UNREACHABLE:       return "SITE_UNREACHABLE";
            case PUNCH_HOLE:             return "PUNCH_HOLE";
            case CLOSE_HOLE:             return "CLOSE_HOLE";
            case GET_VIEW_FROM_COORD:    return "GET_VIEW_FROM_COORD";
            case GET_PING_DATA:          return "GET_PING_DATA";
            case GET_SECRET_KEY:         return "GET_SECRET_KEY";
            case SET_SECRET_KEY:         return "SET_SECRET_KEY";
            case UNLOCK_FORCE:           return "UNLOCK_FORCE";
            case INSTALL_MERGE_VIEW:     return "INSTALL_MERGE_VIEW";

            case USER_DEFINED:           return "USER_DEFINED";
            default:                     return "UNDEFINED(" + t + ")";

        }
    }

    public String toString() {return String.format("%s, arg=%s", type2String(type), arg);}

}
